source/s3_meta_request.c

```diff
@@ -23,6 +23,12 @@ struct s3_meta_request_binding {
     /* Reference to python object that reference to other related python object to keep it alive */
     PyObject *py_core;
 
+    /* If error headers included from finish call, store the python copy of the headers */
+    PyObject *error_headers;
+    /* If error body included from finish result, deep copy of the response body */
+    struct aws_byte_buf error_response_body;
+    int error_code;
```

Comment:
probably simpler to make these changes in python rather than C, since this is a temporary workaround we'll be unwinding in the next few days. You could make the following changes in _S3RequestCore:
```
def _on_finish(self, error_code, error_headers, error_body):
    # temporary hack: delay on_done callback until shutdown
    self._done_error_code = error_code
    self._done_error_headers = error_headers
    self._done_error_body = error_body

    self._owner._binding = None # force S3Request's C parts to be cleaned up
    self._owner = None # allow S3Request to be cleaned up

def _on_shutdown(self):
    ... all the logic that used to be in _on_finish...
    ... the rest of the normal shutdown logic...

```